### ISSUE TITLE
thisType versus this

### ISSUE THREAD

**Issue by electricessence** [Microsoft/TypeScript#9096]
https://github.com/Microsoft/TypeScript/issues/9063#issuecomment-225079380

Wanted to open the discussion to possibly allowing for something other than ```this``` for polymorphic typing.

Right now...
```typescript
class A<T> {
    clone():this {
       return <any> new A<T>();  // without <any> it complains.
    }
   me():this {
      return this; // doesn't complain
   }
}
```

And it would be really nice to be able to:
```typescript
class A<T> {
    map<TResult>(mapper: (v: T) => TResult): this<TResult>{
       // ...
    }
}
```

**Comment by DanielRosenwasser** [comment]
But your `clone` method *doesn't* return something of type `this`, it returns something of type `A<T>`. If you extended `A` in class `B`, clone wouldn't give you something with the members that `B` has.

A simple way to look at it is that the `this` type is supposed to model the type of the expression that you're calling a method off of.

As for the second thing, I think that's a duplicate of #6223.

**Comment by kitsonk** [comment]
It is, and the biggest challenge discussed in there is that if the arity of the generic arguments changes, how is this handled.
